### DISCUSSION TITLE
How to render a window with an OpenGlControl to RenderTargetBitmap?

### DISCUSSION THREAD

**Discussion by marklam** [AvaloniaUI/Avalonia#7772] [Q&A] [answered]
I'm running on Windows with Skia and UseWgl=true.

I have a MainWindow which uses a control derived from `OpenGlControlBase`, and when I try to render to RenderTargetBitmap:

```F#
            use bmp = new RenderTargetBitmap(psize, Vector(96,96))
            this.Measure(size);
            this.Arrange(new Rect(size));
            bmp.Render(this)
```
I get an `ArgumentNullException` under OpenGlBitmapImp.cs line 44
```text
 	SkiaSharp.dll!SkiaSharp.SKSurface.Create(SkiaSharp.GRRecordingContext context, SkiaSharp.GRBackendTexture texture, SkiaSharp.GRSurfaceOrigin origin, int sampleCount, SkiaSharp.SKColorType colorType, SkiaSharp.SKColorSpace colorspace, SkiaSharp.SKSurfaceProperties props)	Unknown
 	SkiaSharp.dll!SkiaSharp.SKSurface.Create(SkiaSharp.GRRecordingContext context, SkiaSharp.GRBackendTexture texture, SkiaSharp.GRSurfaceOrigin origin, SkiaSharp.SKColorType colorType)	Unknown
 	SkiaSharp.dll!SkiaSharp.SKSurface.Create(SkiaSharp.GRContext context, SkiaSharp.GRBackendTexture texture, SkiaSharp.GRSurfaceOrigin origin, SkiaSharp.SKColorType colorType)	Unknown
>	Avalonia.Skia.dll!Avalonia.Skia.GlOpenGlBitmapImpl.Draw(Avalonia.Skia.DrawingContextImpl context, SkiaSharp.SKRect sourceRect, SkiaSharp.SKRect destRect, SkiaSharp.SKPaint paint) Line 44	C#
```
`context.GrContext` is `null` - do I need to do something before I can render the OpenGL control to RenderTargetBitmap?

Thanks for any help.

**Comment by kekekeks** [comment, marked as answer]
Unfortunately right now RenderTargetBitmap is always software-only, so it's not possible to use GPU-backend surfaced to render into it.
We are reworking our rendering stack to make them GPU backed, but that effort is temporally stalled due to recent events.

**Reply by marklam** [reply]
Thanks @kekekeks. Good to know I wasn't just doing it wrong :-) 

For background, I was trying to turn an Avalonia animation into a video, without skipping any frames, by using a custom clock to advance each frame as it was captured into a RenderTargetBitmap and encoded with an ffmpeg wrapper.

**Comment by atticus-lv** [comment]
4 years pass, is there any progress on this issue? @kekekeks
